File: Cpp/Algorithm/src/SpecialSubject/LongestSubarray_1438.cpp
#include "ss.h"
#include <deque>
using namespace myAlgo;
int LongestSubarray_1438::longestSubarray(vector<int>& nums, int limit)
{
	int maxLen = 0;
	deque<int> maxDeque, minDeque;
	for (int l = 0,r=0; r < nums.size(); r++)
	{
		while (!maxDeque.empty() && nums[r] >= nums[maxDeque.back()]) maxDeque.pop_back();
		maxDeque.push_back(r);
		while (!minDeque.empty() && nums[r] <= nums[minDeque.back()]) minDeque.pop_back();
		minDeque.push_back(r);
		while (abs(nums[maxDeque.front()] - nums[minDeque.front()])>limit)
		{
			l++;
			if (maxDeque.front() < l) maxDeque.pop_front();
			if (minDeque.front() < l) minDeque.pop_front();
		}
		maxLen = max(maxLen, r - l + 1);
	}
	return maxLen;
}

File: JavaAlgorithm/src/main/java/algo/tzashinorpu/GSSY/DoublePoint/longestSubarray_1438.java
package algo.tzashinorpu.GSSY.DoublePoint;

import java.util.ArrayDeque;

public class longestSubarray_1438 {
    public int longestSubarray(int[] nums, int limit) {
        int maxLen = 0;
        ArrayDeque<Integer> max = new ArrayDeque<>();
        ArrayDeque<Integer> min = new ArrayDeque<>();
        for (int l = 0, r = 0; r < nums.length; r++) {
            // = 是为了走得更远 [8,8]
            while (!max.isEmpty() && nums[r] >= nums[max.peekLast()]) max.pollLast();
            max.offerLast(r);
            while (!min.isEmpty() && nums[r] <= nums[min.peekLast()]) min.pollLast();
            min.offerLast(r);
            // 移除失效的 最大最小值  滑动窗口
            // [4,8,2,1]  limit=5
            while (Math.abs(nums[max.peekFirst()] - nums[min.peekFirst()]) > limit) {
                l++;
                if (max.peekFirst() < l) max.pollFirst();
                if (min.peekFirst() < l) min.pollFirst();
            }
            maxLen = Math.max(maxLen, r - l + 1);
        }
        return maxLen;
    }
}
